Parameter Interpolatio

proteneer opened the pull request:
This PR sets up parameter interpolation for bond, angle, torsion, nonbonded, chiral_atom/bond terms. It roughly operates in two passes:

1) For a given potential, define its idxs and parameters at the end-states as `idxs0, p0` and `idxs1, p1`
2) `idxs0` and `idxs1` are aligned and deduplicated to generate a list of tuples `idxs_c, (p0, p1)`
3) A simple linear interpolation scheme `p_i(lambda)=(1-lambda)*p0 + lambda*p1` is implemented, but @mcwitt will be optimizing this extensively to be more efficient.

There were lots of tricky cases to cover:

1) for bonded terms that are being turned off,  equilibrium bond lengths, angles, and phases are *not* modified. only the force constants are scaled down.
2) for nonbonded terms that are being turned off, `q_ij`, `sig_ij`, `eps_ij` are *all* scaled down.
3) for chiral bond restraints, the deduplication uses the `sign` information in addition to the `idxs` 
4) for torsions, the deduplication uses the `period` information in addition to the `idxs`

Not done yet:

documentation on tests some of the functions - trying to figure out what level of detail I should add to them.

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -612,3 +612,95 @@ def U_fn(x):
+        """
+        src_system = self.setup_end_state_src()
+        dst_system = self.setup_end_state_dst()
+        interpolate_fn = interpolate.linear_interpolation
```
@mcwitt this is probably where you will supply more efficient interpolation schemes later on.

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -612,3 +612,95 @@ def U_fn(x):
+        torsion = self._setup_intermediate_bonded_term(
+            src_system.torsion, dst_system.torsion, lamb, interpolate.align_torsion_idxs_and_params, interpolate_fn
+        )
+        nonbonded = self._setup_intermediate_bonded_term(
```
@mcwitt you probably want to modify this to do some sort of 4D decoupling scheme (possibly in addition to some kind of parameter interpolation scheme).

proteneer commented on tests/test_interpolate_fe.py:
```diff
@@ -0,0 +1,104 @@
+
+            print(
+                f"BAR: lambda {lambda_schedule[lambda_idx-1]:.3f} -> {lambda_schedule[lambda_idx]:.3f} dG: {dG_exact:.3f} dG_err: {exact_bar_err:.3f}"
+            )
```
note: this code doens't actually test anything - not sure what quantities would be good here... `exact_bar_err not None`?

mcwitt replied:
maybe `np.isfinite` on `dG_exact` and `exact_bar_err`?

proteneer replied:
This is a pretty expensive test to run as well - seems like a bit of a waste just to assert that `dG_exact` is finite

mcwitt commented on tests/test_interpolate.py:
```diff
@@ -0,0 +1,227 @@
+def test_align_harmonic_bond():
+    """
+    Test that we can align idxs and parameters correctly for harmonic bonds.
+    We expect that decoupled terms have their force constants turned set to zero,
```
nit: typo
```suggestion
    We expect that decoupled terms have their force constants set to zero,
```

mcwitt commented on tests/test_interpolate.py:
```diff
@@ -0,0 +1,227 @@
+    a, b, c, d, e, f, g, h, i, j = np.random.rand(10)
+
+    # tbd: what do we do if there are repeats?
+    # merge repeats into a single term first?
```
For now, would it make sense to raise an error if there are repeats? (to prevent any potentially surprising behavior)

proteneer replied:
It currently asserts

mcwitt replied:
> It currently asserts

nit: add test for this?

proteneer replied:
willdo

mcwitt commented on tests/test_interpolate.py:
```diff
@@ -0,0 +1,227 @@
+def test_align_chiral_bonds():
+    """
+    Similar to the chiral_atoms test, except that we check to see if deduplication
+    is using the sign information correctly.
```
nit: might be explicit what "correctly" means here (e.g. as part of the alignment key)

mcwitt commented on timemachine/fe/interpolate.py:
```diff
@@ -0,0 +1,217 @@
+    for k in src_kv.keys():
+        res.add((tuple(k), src_kv[k], dst_kv[k]))
+
+    return res
```
nit: I think this could be simplified a bit (and remove the need for the helper function) as something like:

```python
def align_harmonic_bond_or_angle_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):

    src_kv = dict(zip(src_idxs, src_params))
    dst_kv = dict(zip(dst_idxs, dst_params))

    aligned = {
        (
            k,
            tuple(src_kv[k]) if k in src_kv else (0, dst_kv[k][1]),
            tuple(dst_kv[k]) if k in dst_kv else (0, src_kv[k][1]),
        )
        for k in set(src_kv.keys()).union(dst_kv)
    }

    return aligned
```

mcwitt replied:
I think the various `align_*` functions defined here have a common pattern that might be useful to factor out.

E.g. with a common helper like
```python
def align_idxs_and_params(
    src_idxs,
    src_params,
    dst_idxs,
    dst_params,
    make_default,
    key=lambda idxs, _: idxs,
    get_idxs=lambda key: key,
):
    def to_hashable(x):
        return x if isinstance(x, Hashable) else tuple(x)

    src_kv = {to_hashable(key(idxs, params)): to_hashable(params) for idxs, params in zip(src_idxs, src_params)}
    dst_kv = {to_hashable(key(idxs, params)): to_hashable(params) for idxs, params in zip(dst_idxs, dst_params)}

    return {
        (
            get_idxs(k),
            src_kv[k] if k in src_kv else make_default(dst_kv[k]),
            dst_kv[k] if k in dst_kv else make_default(src_kv[k]),
        )
        for k in set(src_kv.keys()).union(dst_kv)
    }
```,

then the others can become one-liners, e.g.

```python
align_harmonic_bond_or_angle_idxs_and_params = partial(align_idxs_and_params, make_default=lambda p: (0, p[1]))
align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)

align_torsion_idxs_and_params = partial(
    align_idxs_and_params,
    make_default=lambda p: (0, p[1], p[2]),
    key=lambda idxs, params: (idxs, params[2]),
    get_idxs=lambda key: key[0],
)
```

proteneer commented:
After some more debugging, it looks like the nonbonded terms are already numerically unstable for any transformation that involves a dummy hydroxyl that might clash:

There are some absurdly small terms for intermediate windows:

`[i,j] [q_ij, s_ij, e_ij]`
`[ 5 33] [-5.46873960e+00  1.09283576e-01  4.94850147e-03]`

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -612,3 +612,95 @@ def U_fn(x):
+            src_bond.get_signs(),
+            dst_bond.get_idxs(),
+            dst_bond.params,
+            dst_bond.get_idxs(),
```
good call

proteneer commented:
2 changes:

1) modified nonbonded_v3_pairlist_precomputed to take in an offset parameter to implement 4D decoupling
2) modified the interpolation function for nonbonded parameters to instead interpolate the offset from 0 to cutoff (or cutoff to 0).

proteneer commented on timemachine/fe/single_topology_v3.py:
```diff
@@ -612,3 +612,95 @@ def U_fn(x):
+            chiral_bond_params.append(new_params)
+            chiral_bond_signs.append(sign)
+
+        # these should be properly sized
+        chiral_bond_idxs = np.array(chiral_bond_idxs, dtype=np.int32).reshape((-1, 4))
```
the reshape is to deal with weird zero shaped things like an empty list `[]` which need to be re-interpreted as a `(0, 4)` shape in order for the jax functions downstream to be robust.

maxentile commented on tests/test_interpolate.py:
```diff
@@ -0,0 +1,227 @@
+        U_ref = system_dst.get_U_fn()
+        U_test = system_lambda_1.get_U_fn()
+
+        np.testing.assert_almost_equal(U_ref(x0), U_test(x0))
```
>How random are we talking about here? Sampled from MD?

I don't think it has to be sampled from MD, just it looked like the test could easily be made stronger if it looped over a few conformers with different values of the relevant internal d.o.f.s (distances referenced by harmonic bonds, angles referenced by harmonic angles, etc.).

Something like

```python
np.random.seed(2022)
xs = [x0]
for _ in range(10):
   xs.append(x0 + 0.01 * np.random.randn(*x0.shape))

for x in xs:
   np.testing.assert_almost_equal(U_ref(x), U_test(x))
```

would more than suffice

proteneer replied:
ah good call! Thanks that's easy!

proteneer commented on tests/test_interpolate.py:
```diff
@@ -0,0 +1,227 @@
+    We expect that decoupled terms have their force constants turned set to zero,
+    while maintaining the same equilibrium bond lengths.
+    """
+    a, b, c, d, e, f, g, h, i, j = np.random.rand(10)
```
I don't think this really matters that much - this could be literally just `np.arange(10)` if that's more preferable

proteneer replied:
This test is still deterministic in the output even if a random number is used in the input. 

Consider: you can have a function f(x)=0 have random inputs but the output will always be zero. This is a deterministic output.

Furthermore, you can have a function f(x)=x assert that output=input for any value of x (even random ones). The behavior is completely deterministic. This is what the current usage pattern is, ie we're just asserting that the broadcasting is done correctly on some arbitrary values. 

We require determinism tests for nonbonded tests and randomly generated inputs because there are two sources of non-determinism:

The test function f(x) may return different values if we do not use bitwise checks (for the same value of x). 

The test function f(x) may not agree with a reference function g(x) to a sufficient level of precision for different values of x.

mcwitt replied:
I agree that this seems minor, but just wanted to clarify my reasoning for the initial comment.

Deterministic inputs seem preferable to me, even in cases where we are confident that the outcome is regardless deterministic, because
- We don't need to rely on (potentially flawed, especially in more complex cases) reasoning that the result is indeed deterministic. With nondeterministic inputs, the test might seem suspicious to future readers expecting deterministic tests until they've verified this reasoning for themselves
- There doesn't seem to be a benefit to nondeterministic inputs. If we want to test that the outcome is satisfied for varying inputs, we can still do this deterministically by generating a fixed number of random variates, as @maxentile mentioned
